apps/enterprise-knowledge-retrieval/enterprise_knowledge_retrieval.ipynb

```diff
@@ -0,0 +1,1750 @@
+    "        if j == i + int(0.5 * n):\n",
+    "            j = min(i + n, len(tokens))\n",
+    "        yield tokens[i:j]\n",
+    "        i = j\n",
+    "        \n",
+    "def get_unique_id_for_file_chunk(title, chunk_index):\n",
+    "    return str(title+\"-!\"+str(chunk_index))\n",
+    "\n",
+    "def process_file(x,vector_list):\n",
+    "    url = x['url']\n",
+    "    title = x['title']\n",
+    "    file_body_string = x['text']\n",
+    "\n",
+    "    # Clean up the file string by replacing newlines and double spaces and semi-colons\n",
+    "    clean_file_body_string = file_body_string.replace(\"  \", \" \").replace(\"\\n\", \"; \").replace(';',' ')\n",
+    "    #     \n",
+    "        \n",
+    "    \"\"\"Return a list of tuples (text_chunk, embedding) and an average embedding for a text.\"\"\"\n",
+    "    token_chunks = list(chunks(clean_file_body_string, TEXT_EMBEDDING_CHUNK_SIZE, tokenizer))\n",
+    "    text_chunks = [f'Title: {title};\\n'+ tokenizer.decode(chunk) for chunk in token_chunks]\n",
+    "    \n",
+    "    embeddings_response = openai.Embedding.create(input=text_chunks, model=EMBEDDINGS_MODEL)\n",
+    "\n",
+    "    embeddings = [embedding[\"embedding\"] for embedding in embeddings_response['data']]\n",
+    "    text_embeddings = list(zip(text_chunks, embeddings))\n",
+    "\n",
+    "    # Get the vectors array of triples: file_chunk_id, embedding, metadata for each embedding\n",
+    "    # Metadata is a dict with keys: filename, file_chunk_index\n",
+    "    \n",
+    "    for i, (text_chunk, embedding) in enumerate(text_embeddings):\n",
+    "        id = get_unique_id_for_file_chunk(title, i)\n",
+    "        vector_list.append(({'id': id\n",
+    "                         , \"vector\": embedding, 'metadata': {\"url\": x['url']\n",
+    "                                                              ,\"title\": title\n",
+    "                                                              , \"content\": text_chunk\n",
+    "                                                              , \"file_chunk_index\": i}}))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 130,
+   "id": "dfeff174",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 12.5 s, sys: 2.48 s, total: 15 s\n",
+      "Wall time: 11min 47s\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "0       None\n",
+       "1       None\n",
+       "2       None\n",
+       "3       None\n",
+       "4       None\n",
+       "        ... \n",
+       "1995    None\n",
+       "1996    None\n",
+       "1997    None\n",
+       "1998    None\n",
+       "1999    None\n",
+       "Length: 2000, dtype: object"
+      ]
+     },
+     "execution_count": 130,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "# This step takes about 13 minutes\n",
```

Comment:
13 minutes is long enough that it could be worth calling in batches or parallelizing. Or at least adding a note that it's an option. But not high priority.